File: Parallel-Csharp/src/Parallel/Linq101/MyLinq/WhereMethods.cs
using Parallel.Linq101.DataSources;

namespace Parallel.Linq101.MyLinq; 

public class WhereMethods {
    public List<Product> GetProductList() => Products.ProductList;
    public List<Customer> GetCustomerList() => Customers.CustomerList;
    
    //original:5 target:10 just indexed where form Restrictions
    public List<int> IntIndexedWhere() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> q = numbers.Where((n,i)=>n<i).ToList();
        return q;
    }
    
    public List<int> OddIntIndexedWhere() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> q = numbers.Where(i => i % 2 == 1).Where((n,i)=>n<i).ToList();
        return q;
    }
    
    public List<int> OddIntOriginalIndexedWhere() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> q = numbers.Where((n,i)=>n<i&&n % 2 == 1).ToList();
        return q;
    }
    public List<int> IntSquareIndexedWhere() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> q = numbers.Select(i => i * i).Where((n,i)=>n<i).ToList();
        return q;
    }
    
    public List<int> IntWithSquareIndexedWhere() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        List<int> q = numbers.Where((n,i)=>n<i*i).ToList();
        return q;
    }

    public List<string> StringIndexedWhere() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        List<string> q = words.Where((w,i)=>w.Length<i).ToList();
        return q;
    }

    public List<string> LongStringIndexedWhere() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        List<string> q =words.Where(w=>w.Length>5).Where((w,i)=>w.Length<i*3).ToList();
        return q;
    }

    public List<string> LongStringOrginalIndexedWhere() {
        string[] words = { "cherry", "apple", "blueberry", "banana", "abacus" };
        List<string> q = words.Where((w,i)=>w.Length<i*3&&w.Length>5).ToList();
        return q;
    }
    
    public List<Product> ProductIndexedWhere() {
        List<Product> products = GetProductList();
        List<Product> q = products.Where((w,i)=>w.UnitsInStock>0&&i%2==1).ToList();
        return q;
    }
    
    public List<Product> ExpensiveProductIndexedWhere() {
        List<Product> products = GetProductList();
        List<Product> q = products.Where((w,i)=>w.UnitPrice>100&&i%2==1).ToList();
        return q;
    }
}

File: Parallel-Java/src/parallel/linq101/mylinq/WhereMethods.java
package parallel.linq101.mylinq;

import parallel.linq101.datasource.Customer;
import parallel.linq101.datasource.Data;
import parallel.linq101.datasource.Product;

import java.util.Arrays;
import java.util.List;
import java.util.stream.IntStream;

public class WhereMethods {
    public List<Product> getProductList() {
        return Data.getProductList();
    }

    public List<Customer> getCustomerList() {
        return Data.getCustomerList();
    }

    public List<Integer> intIndexedWhere() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> q = IntStream.range(0, numbers.length).filter(i -> numbers[i] < i).map(i -> numbers[i]).boxed().toList();
        return q;
    }

    public List<Integer> oddIntIndexedWhere() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int[] oddNumbers = Arrays.stream(numbers).filter(i -> i % 2 == 1).toArray();
        List<Integer> q = IntStream.range(0, oddNumbers.length).filter(i -> oddNumbers[i] < i)
                .map(i -> oddNumbers[i]).boxed().toList();
        return q;
    }

    public List<Integer> oddIntOriginalIndexedWhere() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> q = IntStream.range(0, numbers.length).filter(i -> numbers[i] < i && numbers[i] % 2 == 1)
                .map(i -> numbers[i]).boxed().toList();
        return q;
    }

    public List<Integer> intSquareIndexedWhere() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> q = IntStream.range(0, numbers.length).filter(i -> numbers[i] * numbers[i] < i)
                .map(i -> numbers[i] * numbers[i]).boxed().toList();
        return q;
    }

    public List<Integer> intWithSquareIndexedWhere() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        List<Integer> q = IntStream.range(0, numbers.length).filter(i -> numbers[i] < i * i)
                .map(i -> numbers[i]).boxed().toList();
        return q;
    }

    public List<String> stringIndexedWhere() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        IntStream.range(0, words.length).filter(i -> words[i].length() < i).mapToObj(i -> words[i]).toList();
        List<String> q = IntStream.range(0, words.length).filter(i -> words[i].length() < i).mapToObj(i -> words[i]).toList();
        return q;
    }

    public List<String> longStringIndexedWhere() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        String[] longWords = (String[]) Arrays.stream(words).filter(w -> w.length() > 5).toArray();
        List<String> q = IntStream.range(0, longWords.length).filter(i -> longWords[i].length() < i * 3)
                .mapToObj(i -> longWords[i]).toList();
        return q;
    }

    public List<String> longStringOrginalIndexedWhere() {
        String[] words = {"cherry", "apple", "blueberry", "banana", "abacus"};
        List<String> q = IntStream.range(0, words.length).filter(i -> words[i].length() < i * 3 &&
                words[i].length() > 5).mapToObj(i -> words[i]).toList();
        return q;
    }

    public List<Product> productIndexedWhere() {
        List<Product> products = getProductList();
        List<Product> q = IntStream.range(0, products.size()).filter(i -> products.get(i).getUnitsInStock() > 0 &&
                i % 2 == 1).mapToObj(i -> products.get(i)).toList();
        return q;
    }

    public List<Product> expensiveProductIndexedWhere() {
        List<Product> products = getProductList();
        List<Product> q = IntStream.range(0, products.size()).filter(i -> products.get(i).getUnitPrice() > 100 &&
                i % 2 == 1).mapToObj(i -> products.get(i)).toList();
        return q;
    }
}
